/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package Fractale;

import javax.swing.JFileChooser;

/** This class allows the display of fractal graphics using a dedicated JPanel (see MyFractalPanel)
 *
 * @author yohann.lelievre
 */
public class AffichageFractale extends javax.swing.JFrame {

    /**
     * Creates new form AffichageFractale
     */
    public AffichageFractale() {
        initComponents();
        startBase       = 1;            // Début de la séquence ADN lue
        nbSequence      = 100000;       // Longueur de la séquence ADN lue
        nomFichier      = "";           // Nom du fichier et chemin d'accès de la séquence ADN étudiée
        pas             = 100;          // Pas de lecture entre 2 déplacement de la scroll bar
        typeFractale    = 1;            // Type d'analyse facale (Par défaut: analyse en noir et blanc)
    }

    /**
     * This method is called from within the constructor to initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is always
     * regenerated by the Form Editor.
     */
    @SuppressWarnings("unchecked")
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    private void initComponents() {

        fichierADNFlChsr = new javax.swing.JFileChooser();
        titreLbl = new javax.swing.JLabel();
        selectionFichierLbl = new javax.swing.JLabel();
        fichierBtn = new javax.swing.JButton();
        fractalePnl = new MyFractalPanel();
        basesLbl = new javax.swing.JLabel();
        ADNScrlBr = new javax.swing.JScrollBar();
        nbBasesLbl = new javax.swing.JLabel();
        nbSequenceLbl = new javax.swing.JLabel();
        nbSequenceTxtFld = new javax.swing.JTextField();
        debutSequenceLbl = new javax.swing.JLabel();
        debutSequenceTxtFld = new javax.swing.JTextField();
        typeFractalLbl = new javax.swing.JLabel();
        typeFractalCmbBx = new javax.swing.JComboBox();
        CLbl = new javax.swing.JLabel();
        GLbl = new javax.swing.JLabel();
        ALbl = new javax.swing.JLabel();
        TLbl = new javax.swing.JLabel();
        pasLbl = new javax.swing.JLabel();
        pasTxtFld = new javax.swing.JTextField();
        dimensionFractaleLbl = new javax.swing.JLabel();
        dimensionFractaleTxtFld = new javax.swing.JTextField();

        fichierADNFlChsr.setApproveButtonToolTipText("");
        fichierADNFlChsr.setCurrentDirectory(new java.io.File("C:\\Users\\yohann.lelievre\\Documents\\NetBeansProjects\\Fractale\\FractaleFolies\\genome\\hs_alt_CHM1_1.1_chr14.fa"));

        setDefaultCloseOperation(javax.swing.WindowConstants.EXIT_ON_CLOSE);
        addFocusListener(new java.awt.event.FocusAdapter() {
            public void focusLost(java.awt.event.FocusEvent evt) {
                formFocusLost(evt);
            }
        });

        titreLbl.setFont(new java.awt.Font("Bookman Old Style", 2, 24)); // NOI18N
        titreLbl.setHorizontalAlignment(javax.swing.SwingConstants.CENTER);
        titreLbl.setText("Balade Fractale au coeur de l'ADN");

        selectionFichierLbl.setBackground(new java.awt.Color(255, 255, 255));
        selectionFichierLbl.setText("Sélection du fichier FastA");

        fichierBtn.setText("Fichier ADN");
        fichierBtn.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                fichierBtnActionPerformed(evt);
            }
        });

        fractalePnl.setBackground(new java.awt.Color(255, 255, 255));

        javax.swing.GroupLayout fractalePnlLayout = new javax.swing.GroupLayout(fractalePnl);
        fractalePnl.setLayout(fractalePnlLayout);
        fractalePnlLayout.setHorizontalGroup(
            fractalePnlLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGap(0, 512, Short.MAX_VALUE)
        );
        fractalePnlLayout.setVerticalGroup(
            fractalePnlLayout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGap(0, 512, Short.MAX_VALUE)
        );

        basesLbl.setText("Nombre de bases du fichier ADN:");

        ADNScrlBr.setBlockIncrement(1);
        ADNScrlBr.setMaximum(1);
        ADNScrlBr.setMinimum(1);
        ADNScrlBr.setOrientation(javax.swing.JScrollBar.HORIZONTAL);
        ADNScrlBr.setToolTipText("");
        ADNScrlBr.setUnitIncrement(100);
        ADNScrlBr.addAdjustmentListener(new java.awt.event.AdjustmentListener() {
            public void adjustmentValueChanged(java.awt.event.AdjustmentEvent evt) {
                ADNScrlBrAdjustmentValueChanged(evt);
            }
        });

        nbBasesLbl.setHorizontalAlignment(javax.swing.SwingConstants.TRAILING);
        nbBasesLbl.setText("0");
        nbBasesLbl.setHorizontalTextPosition(javax.swing.SwingConstants.RIGHT);

        nbSequenceLbl.setText("Longueur de la séquence analysée:");

        nbSequenceTxtFld.setHorizontalAlignment(javax.swing.JTextField.TRAILING);
        nbSequenceTxtFld.setText("100000");
        nbSequenceTxtFld.addFocusListener(new java.awt.event.FocusAdapter() {
            public void focusLost(java.awt.event.FocusEvent evt) {
                nbSequenceTxtFldFocusLost(evt);
            }
        });
        nbSequenceTxtFld.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                nbSequenceTxtFldActionPerformed(evt);
            }
        });

        debutSequenceLbl.setText("Début de la séquence analysée:");

        debutSequenceTxtFld.setHorizontalAlignment(javax.swing.JTextField.TRAILING);
        debutSequenceTxtFld.setText("1");
        debutSequenceTxtFld.setEnabled(false);
        debutSequenceTxtFld.addFocusListener(new java.awt.event.FocusAdapter() {
            public void focusLost(java.awt.event.FocusEvent evt) {
                debutSequenceTxtFldFocusLost(evt);
            }
        });
        debutSequenceTxtFld.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                debutSequenceTxtFldActionPerformed(evt);
            }
        });

        typeFractalLbl.setText("Type de représentation fractale");

        typeFractalCmbBx.setModel(new javax.swing.DefaultComboBoxModel(new String[] { "Noir et Blanc", "Nuance de Gris", "Couleur" }));
        typeFractalCmbBx.setToolTipText("");
        typeFractalCmbBx.addItemListener(new java.awt.event.ItemListener() {
            public void itemStateChanged(java.awt.event.ItemEvent evt) {
                typeFractalCmbBxItemStateChanged(evt);
            }
        });
        typeFractalCmbBx.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                typeFractalCmbBxActionPerformed(evt);
            }
        });

        CLbl.setFont(new java.awt.Font("Tahoma", 1, 18)); // NOI18N
        CLbl.setText("C");

        GLbl.setFont(new java.awt.Font("Tahoma", 1, 18)); // NOI18N
        GLbl.setText("G");

        ALbl.setFont(new java.awt.Font("Tahoma", 1, 18)); // NOI18N
        ALbl.setText("A");

        TLbl.setFont(new java.awt.Font("Tahoma", 1, 18)); // NOI18N
        TLbl.setText("T");

        pasLbl.setText("Pas de parcours");

        pasTxtFld.setHorizontalAlignment(javax.swing.JTextField.TRAILING);
        pasTxtFld.setText("100");
        pasTxtFld.addFocusListener(new java.awt.event.FocusAdapter() {
            public void focusLost(java.awt.event.FocusEvent evt) {
                pasTxtFldFocusLost(evt);
            }
        });
        pasTxtFld.addActionListener(new java.awt.event.ActionListener() {
            public void actionPerformed(java.awt.event.ActionEvent evt) {
                pasTxtFldActionPerformed(evt);
            }
        });

        dimensionFractaleLbl.setText("Dimension fractale apparante (Minkowski–Bouligand):");

        dimensionFractaleTxtFld.setHorizontalAlignment(javax.swing.JTextField.TRAILING);
        dimensionFractaleTxtFld.setToolTipText("");
        dimensionFractaleTxtFld.setEnabled(false);

        javax.swing.GroupLayout layout = new javax.swing.GroupLayout(getContentPane());
        getContentPane().setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addContainerGap()
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                    .addComponent(titreLbl, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addGroup(layout.createSequentialGroup()
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addComponent(CLbl)
                            .addComponent(ALbl))
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addComponent(ADNScrlBr, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                            .addComponent(fractalePnl, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(GLbl)
                        .addGap(21, 21, 21)
                        .addComponent(fichierBtn))
                    .addComponent(TLbl))
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addComponent(selectionFichierLbl, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                    .addGroup(layout.createSequentialGroup()
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                            .addGroup(layout.createSequentialGroup()
                                .addComponent(dimensionFractaleLbl)
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addComponent(dimensionFractaleTxtFld, javax.swing.GroupLayout.PREFERRED_SIZE, 57, javax.swing.GroupLayout.PREFERRED_SIZE))
                            .addGroup(layout.createSequentialGroup()
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                    .addComponent(nbSequenceLbl)
                                    .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                        .addComponent(basesLbl, javax.swing.GroupLayout.Alignment.TRAILING)
                                        .addComponent(debutSequenceLbl)
                                        .addComponent(typeFractalLbl)
                                        .addComponent(pasLbl)))
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                    .addGroup(layout.createSequentialGroup()
                                        .addGap(27, 27, 27)
                                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                            .addComponent(nbBasesLbl, javax.swing.GroupLayout.Alignment.TRAILING, javax.swing.GroupLayout.PREFERRED_SIZE, 119, javax.swing.GroupLayout.PREFERRED_SIZE)
                                            .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                                                .addComponent(debutSequenceTxtFld, javax.swing.GroupLayout.Alignment.TRAILING, javax.swing.GroupLayout.PREFERRED_SIZE, 119, javax.swing.GroupLayout.PREFERRED_SIZE)
                                                .addComponent(nbSequenceTxtFld, javax.swing.GroupLayout.Alignment.TRAILING, javax.swing.GroupLayout.PREFERRED_SIZE, 119, javax.swing.GroupLayout.PREFERRED_SIZE))))
                                    .addGroup(javax.swing.GroupLayout.Alignment.TRAILING, layout.createSequentialGroup()
                                        .addGap(13, 13, 13)
                                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING, false)
                                            .addComponent(typeFractalCmbBx, 0, 119, Short.MAX_VALUE)
                                            .addComponent(pasTxtFld))))))
                        .addContainerGap(257, Short.MAX_VALUE))))
        );
        layout.setVerticalGroup(
            layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
            .addGroup(layout.createSequentialGroup()
                .addGap(18, 18, 18)
                .addComponent(titreLbl, javax.swing.GroupLayout.PREFERRED_SIZE, 24, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                    .addGroup(layout.createSequentialGroup()
                        .addComponent(CLbl)
                        .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                        .addComponent(ALbl))
                    .addComponent(fractalePnl, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE)
                    .addGroup(layout.createSequentialGroup()
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                .addComponent(fichierBtn)
                                .addComponent(selectionFichierLbl, javax.swing.GroupLayout.PREFERRED_SIZE, 23, javax.swing.GroupLayout.PREFERRED_SIZE))
                            .addComponent(GLbl))
                        .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.LEADING)
                            .addGroup(layout.createSequentialGroup()
                                .addGap(0, 0, Short.MAX_VALUE)
                                .addComponent(TLbl))
                            .addGroup(layout.createSequentialGroup()
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                    .addComponent(basesLbl)
                                    .addComponent(nbBasesLbl, javax.swing.GroupLayout.PREFERRED_SIZE, 14, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.TRAILING)
                                    .addComponent(nbSequenceLbl)
                                    .addComponent(nbSequenceTxtFld, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                    .addComponent(debutSequenceLbl)
                                    .addComponent(debutSequenceTxtFld, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.UNRELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                    .addComponent(pasLbl)
                                    .addComponent(pasTxtFld, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                    .addComponent(typeFractalLbl)
                                    .addComponent(typeFractalCmbBx, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))
                                .addPreferredGap(javax.swing.LayoutStyle.ComponentPlacement.RELATED, javax.swing.GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                                .addGroup(layout.createParallelGroup(javax.swing.GroupLayout.Alignment.BASELINE)
                                    .addComponent(dimensionFractaleLbl)
                                    .addComponent(dimensionFractaleTxtFld, javax.swing.GroupLayout.PREFERRED_SIZE, javax.swing.GroupLayout.DEFAULT_SIZE, javax.swing.GroupLayout.PREFERRED_SIZE))))))
                .addGap(18, 18, 18)
                .addComponent(ADNScrlBr, javax.swing.GroupLayout.PREFERRED_SIZE, 17, javax.swing.GroupLayout.PREFERRED_SIZE)
                .addGap(46, 46, 46))
        );

        pack();
    }// </editor-fold>//GEN-END:initComponents

    
    /**
     * Method of AffichageFractale
     * When the button fichierBtn is perfomed, a file chooser dialog box is opened.
     * If the file is valid, the fractal and its dimension is generated and displayed
     */
    private void fichierBtnActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_fichierBtnActionPerformed
        int o = fichierADNFlChsr.showOpenDialog(this);
        if (o == JFileChooser.APPROVE_OPTION){
            java.io.File f = fichierADNFlChsr.getSelectedFile();
            nomFichier = f.getAbsolutePath();
            selectionFichierLbl.setText(nomFichier);
            nbSequence = Integer.parseInt(nbSequenceTxtFld.getText());
            startBase = 1;
            fractale = new ObjetFractale(nomFichier);
            nbBases = fractale.getNombreBase();
            ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
            dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
            fractalePnl.repaint();
            nbBasesLbl.setText(""+nbBases);
            ADNScrlBr.setMaximum((nbBases-nbSequence+1));
            ADNScrlBr.setValue(startBase);
            debutSequenceTxtFld.setEnabled(true);
        }
    }//GEN-LAST:event_fichierBtnActionPerformed

    /**
     * Method of AffichageFractale
     * When the JFrame is perfomed, the new nbSequence is taken into account.
     * If nbSequence is valid, the corresponding fractal and its dimension is generated and displayed
     */
    private void nbSequenceTxtFldActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_nbSequenceTxtFldActionPerformed
        try {
            nbSequence = Integer.parseInt(nbSequenceTxtFld.getText());
            // nbSequence doit être supérieur à 8 car l'analyse fractal se fait sur des mots de 8 bases ADN
            if (nbSequence < 8){
                javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier supérieur ou égal à 8.");
                setVisible(true);
                nbSequenceTxtFld.requestFocus();
            }else{
                // Si le pas est supérieur à nbSequence on positionne le pas au niveau de nbSequence
                if (nbSequence < pas){
                    pas = nbSequence;
                    pasTxtFld.setText(""+pas);
                }
                // Si un fichier est en cours de lecture on réactualise l'affichage de la fractale et de sa dimension
                if (!nomFichier.equals("")) {
                    if (ADNScrlBr.getValue() > (nbBases-nbSequence+1)){
                        ADNScrlBr.setValue((nbBases-nbSequence+1));
                    }
                    ADNScrlBr.setMaximum((nbBases-nbSequence+1));
                    ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
                    dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
                    fractalePnl.repaint();
                }
            }

        } catch (NumberFormatException e){
            javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier supérieur ou égal à 8.");
            setVisible(true);
            nbSequenceTxtFld.requestFocus();
        }
    }//GEN-LAST:event_nbSequenceTxtFldActionPerformed

    private void formFocusLost(java.awt.event.FocusEvent evt) {//GEN-FIRST:event_formFocusLost
    }//GEN-LAST:event_formFocusLost

    /**
     * Method of AffichageFractale
     * When the JFrame is perfomed, the new debutSequence is taken into account.
     * If debutSequence is valid, the corresponding fractal and its dimension is generated and displayed
     */
    private void debutSequenceTxtFldActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_debutSequenceTxtFldActionPerformed
        try {
            int test = Integer.parseInt(debutSequenceTxtFld.getText());
            // Le début de la séquence doit être inférieur à la taille du fichier moins la longueur de la séquence analysée.
            if (test > (nbBases-nbSequence+1)){
                javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif inférieur ou égal à "+(nbBases-nbSequence+1)+".");
                setVisible(true);
                debutSequenceTxtFld.requestFocus();
            }else{
                startBase = test;
                ADNScrlBr.setValue(startBase);
                if (!nomFichier.equals("")) {
                    ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
                    dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
                    fractalePnl.repaint();
                }
            }

        } catch (NumberFormatException e){
            javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif.");
            setVisible(true);
            debutSequenceTxtFld.requestFocus();
        }
    }//GEN-LAST:event_debutSequenceTxtFldActionPerformed

    /**
     * Method of AffichageFractale
     * When the Sroll Bar ADNScrlBr is moved, the new debutSequence is taken into account.
     * The corresponding fractal and its dimension is generated and displayed
     */
    private void ADNScrlBrAdjustmentValueChanged(java.awt.event.AdjustmentEvent evt) {//GEN-FIRST:event_ADNScrlBrAdjustmentValueChanged
        startBase = ADNScrlBr.getValue();
        debutSequenceTxtFld.setText(""+startBase);
        if (!nomFichier.equals("")) {
            ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
            dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
            fractalePnl.repaint();
        }
    }//GEN-LAST:event_ADNScrlBrAdjustmentValueChanged

    /**
     * Method of AffichageFractale
     * When the Text Field nbSequence lost the focus, the new nbSequence is taken into account.
     * If nbSequence is valid, the corresponding fractal and its dimension is generated and displayed
     */
    private void nbSequenceTxtFldFocusLost(java.awt.event.FocusEvent evt) {//GEN-FIRST:event_nbSequenceTxtFldFocusLost
        try {
            nbSequence = Integer.parseInt(nbSequenceTxtFld.getText());
            // nbSequence doit être supérieur à 8 car l'analyse fractal se fait sur des mots de 8 bases ADN
            if (nbSequence < 8){
                javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier supérieur ou égal à 8.");
                setVisible(true);
                nbSequenceTxtFld.requestFocus();
            }else{
                // Si le pas est supérieur à nbSequence on positionne le pas au niveau de nbSequence
                if (nbSequence < pas){
                    pas = nbSequence;
                    pasTxtFld.setText(""+pas);
                }
                // Si un fichier est en cours de lecture on réactualise l'affichage de la fractale et de sa dimension
                if (!nomFichier.equals("")) {
                    if (ADNScrlBr.getValue() > (nbBases-nbSequence+1)){
                        ADNScrlBr.setValue((nbBases-nbSequence+1));
                    }
                    ADNScrlBr.setMaximum((nbBases-nbSequence+1));
                    ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
                    dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
                    fractalePnl.repaint();
                }
            }

        } catch (NumberFormatException e){
            javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier supérieur ou égal à 8.");
            setVisible(true);
            nbSequenceTxtFld.requestFocus();
        }
    }//GEN-LAST:event_nbSequenceTxtFldFocusLost

    /**
     * Method of AffichageFractale
     * When the Text Field debutSequence lost the focus, the new debutSequence is taken into account.
     * If debutSequence is valid, the corresponding fractal and its dimension is generated and displayed
     */
    private void debutSequenceTxtFldFocusLost(java.awt.event.FocusEvent evt) {//GEN-FIRST:event_debutSequenceTxtFldFocusLost
        try {
            int test = Integer.parseInt(debutSequenceTxtFld.getText());
            // Le début de la séquence doit être inférieur à la taille du fichier moins la longueur de la séquence analysée.
            if (test > (nbBases-nbSequence+1)){
                javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif inférieur ou égal à "+(nbBases-nbSequence+1)+".");
                setVisible(true);
                debutSequenceTxtFld.requestFocus();
            }else{
                startBase = test;
                ADNScrlBr.setValue(startBase);
                if (!nomFichier.equals("")) {
                    ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
                    dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
                    fractalePnl.repaint();
                }
            }

        } catch (NumberFormatException e){
            javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif.");
            setVisible(true);
            debutSequenceTxtFld.requestFocus();
        }
    }//GEN-LAST:event_debutSequenceTxtFldFocusLost

    /**
     * Method of AffichageFractale
     * When the JFrame is perfomed, the new typeFractal is taken into account.
     * The corresponding type of fractal and its dimension is generated and displayed
     */
    private void typeFractalCmbBxActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_typeFractalCmbBxActionPerformed
        
        typeFractale = typeFractalCmbBx.getSelectedIndex() + 1;
        if (!nomFichier.equals("")) {
            ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
            dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
            fractalePnl.repaint();
        }
    }//GEN-LAST:event_typeFractalCmbBxActionPerformed

    /**
     * Method of AffichageFractale
     * When the state of the Combo Box typeFractal is changed, the new typeFractal is taken into account.
     * The corresponding type of fractal and its dimension is generated and displayed
     */
    private void typeFractalCmbBxItemStateChanged(java.awt.event.ItemEvent evt) {//GEN-FIRST:event_typeFractalCmbBxItemStateChanged
        
        typeFractale = typeFractalCmbBx.getSelectedIndex() + 1;
        if (!nomFichier.equals("")) {
            ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
            dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
            fractalePnl.repaint();
        }
    }//GEN-LAST:event_typeFractalCmbBxItemStateChanged

    /**
     * Method of AffichageFractale
     * When the JFrame is perfomed, the new pas is taken into account.
     * If pas is valid, the corresponding fractal and its dimension is generated and displayed
     */
    private void pasTxtFldActionPerformed(java.awt.event.ActionEvent evt) {//GEN-FIRST:event_pasTxtFldActionPerformed
        try {
            pas = Integer.parseInt(pasTxtFld.getText());
            // Le pas doit être supérieur ou égal à 8 car l'analyse fractal se fait sur des mots de 8 bases ADN.
            // Le pas doit également être inférieur à la taille de la séquence étudiée.
            if ((pas < 8)||(pas > nbSequence)){
                javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif supérieur à 7 et inférieur à "+nbSequence+".");
                setVisible(true);
                debutSequenceTxtFld.requestFocus();
            }else{
                ADNScrlBr.setUnitIncrement(pas);
                if (!nomFichier.equals("")) {
                    ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
                    dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
                    fractalePnl.repaint();
                }
            }

        } catch (NumberFormatException e){
            javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif.");
            setVisible(true);
            pasTxtFld.requestFocus();
        }
    }//GEN-LAST:event_pasTxtFldActionPerformed

    private void pasTxtFldFocusLost(java.awt.event.FocusEvent evt) {//GEN-FIRST:event_pasTxtFldFocusLost
        try {
            pas = Integer.parseInt(pasTxtFld.getText());
            // Le pas doit être supérieur ou égal à 8 car l'analyse fractal se fait sur des mots de 8 bases ADN.
            // Le pas doit également être inférieur à la taille de la séquence étudiée.
            if ((pas < 8)||(pas > nbSequence)){
                javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif supérieur à 7 et inférieur à "+nbSequence+".");
                setVisible(true);
                debutSequenceTxtFld.requestFocus();
            }else{
                ADNScrlBr.setUnitIncrement(pas);
                if (!nomFichier.equals("")) {
                    ((MyFractalPanel) fractalePnl).calculFractale(fractale, (startBase-1), (startBase+nbSequence-1), typeFractale);
                    dimensionFractaleTxtFld.setText(String.format("%.3f",fractale.getDimensionFractale()));
                    fractalePnl.repaint();
                }
            }

        } catch (NumberFormatException e){
            javax.swing.JOptionPane.showMessageDialog(null,"Veuillez saisir un entier positif.");
            setVisible(true);
            pasTxtFld.requestFocus();
        }
    }//GEN-LAST:event_pasTxtFldFocusLost

    
    /**
     * @param args the command line arguments
     */
    public static void main(String args[]) {
        /* Set the Nimbus look and feel */
        //<editor-fold defaultstate="collapsed" desc=" Look and feel setting code (optional) ">
        /* If Nimbus (introduced in Java SE 6) is not available, stay with the default look and feel.
         * For details see http://download.oracle.com/javase/tutorial/uiswing/lookandfeel/plaf.html 
         */
        try {
            for (javax.swing.UIManager.LookAndFeelInfo info : javax.swing.UIManager.getInstalledLookAndFeels()) {
                if ("Nimbus".equals(info.getName())) {
                    javax.swing.UIManager.setLookAndFeel(info.getClassName());
                    break;
                }
            }
        } catch (ClassNotFoundException ex) {
            java.util.logging.Logger.getLogger(AffichageFractale.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        } catch (InstantiationException ex) {
            java.util.logging.Logger.getLogger(AffichageFractale.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        } catch (IllegalAccessException ex) {
            java.util.logging.Logger.getLogger(AffichageFractale.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        } catch (javax.swing.UnsupportedLookAndFeelException ex) {
            java.util.logging.Logger.getLogger(AffichageFractale.class.getName()).log(java.util.logging.Level.SEVERE, null, ex);
        }
        //</editor-fold>

        /* Create and display the form */
        java.awt.EventQueue.invokeLater(new Runnable() {
            public void run() {
                new AffichageFractale().setVisible(true);
            }
        });
    }
    
    private String nomFichier;
    private int startBase;
    private int nbSequence;
    private int nbBases;
    private int pas;
    private int typeFractale;
    private ObjetFractale fractale;
    
    // Variables declaration - do not modify//GEN-BEGIN:variables
    private javax.swing.JScrollBar ADNScrlBr;
    private javax.swing.JLabel ALbl;
    private javax.swing.JLabel CLbl;
    private javax.swing.JLabel GLbl;
    private javax.swing.JLabel TLbl;
    private javax.swing.JLabel basesLbl;
    private javax.swing.JLabel debutSequenceLbl;
    private javax.swing.JTextField debutSequenceTxtFld;
    private javax.swing.JLabel dimensionFractaleLbl;
    private javax.swing.JTextField dimensionFractaleTxtFld;
    private javax.swing.JFileChooser fichierADNFlChsr;
    private javax.swing.JButton fichierBtn;
    private javax.swing.JPanel fractalePnl;
    private javax.swing.JLabel nbBasesLbl;
    private javax.swing.JLabel nbSequenceLbl;
    private javax.swing.JTextField nbSequenceTxtFld;
    private javax.swing.JLabel pasLbl;
    private javax.swing.JTextField pasTxtFld;
    private javax.swing.JLabel selectionFichierLbl;
    private javax.swing.JLabel titreLbl;
    private javax.swing.JComboBox typeFractalCmbBx;
    private javax.swing.JLabel typeFractalLbl;
    // End of variables declaration//GEN-END:variables
}
